### DISCUSSION TITLE
Timer Interrupt

### DISCUSSION THREAD

**Discussion by ennio64** [stm32duino/Arduino_Core_STM32#2170] [Q&A] [answered]
#include "stm32f4xx_hal.h"

TIM_HandleTypeDef htim3;
uint32_t interrupt_count = 0;

void setup() {
  pinMode(LED_BUILTIN, OUTPUT);
  digitalWrite(LED_BUILTIN, HIGH);

  HAL_Init();
  __enable_irq();
  
  __HAL_RCC_TIM3_CLK_ENABLE();

  // Configurazione del Timer 3
  htim3.Instance = TIM3;
  htim3.Init.Prescaler = 84000 - 1;
  htim3.Init.CounterMode = TIM_COUNTERMODE_UP;
  htim3.Init.Period = 6000;  
  htim3.Init.ClockDivision = TIM_CLOCKDIVISION_DIV1;

  HAL_TIM_Base_DeInit(&htim3);
  HAL_TIM_Base_Init(&htim3);
 
  __HAL_TIM_ENABLE_IT(&htim3, TIM_IT_UPDATE); 

  //HAL_NVIC_SetPriority(TIM3_IRQn, 0, 1);
  // HAL_NVIC_EnableIRQ(TIM3_IRQn);  

  // Avvia il Timer 3
   HAL_TIM_Base_Start_IT(&htim3);
}

void loop() {
}

void TIM3_IRQHandler(void) {
  if (__HAL_TIM_GET_FLAG(&htim3, TIM_FLAG_UPDATE) != RESET && __HAL_TIM_GET_IT_SOURCE(&htim3, TIM_IT_UPDATE) != RESET) {
    __HAL_TIM_CLEAR_IT(&htim3, TIM_IT_UPDATE);
    digitalWrite(LED_BUILTIN, !digitalRead(LED_BUILTIN));
    interrupt_count++; 
  }
}
This code is correctly compiled but the interrupt is not generated.

If I change void TIM3_IRQHandler(void) with

void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim) {
  if (htim->Instance == TIM3) {
    // Gestisci l'interrupt del timer TIM3 qui
    digitalWrite(LED_BUILTIN, digitalRead(!LED_BUILTIN));
    interrupt_count++;
  }
}
Compilation error:
c:/users/myself/appdata/local/arduino15/packages/stmicroelectronics/tools/xpack-arm-none-eabi-gcc/12.2.1-1.2/bin/../lib/gcc/arm-none-eabi/12.2.1/../../../../arm-none-eabi/bin/ld.exe: C:\Users\MYSELF\AppData\Local\Temp\arduino\sketches\D5AFF887F9B4FFAFCC5B5151A01F0E99\libraries\SrcWrapper\HardwareTimer.cpp.o: in function `HAL_TIM_PeriodElapsedCallback':
HardwareTimer.cpp:(.text.HAL_TIM_PeriodElapsedCallback+0x0): multiple definition of `HAL_TIM_PeriodElapsedCallback'; C:\Users\MYSELF\AppData\Local\Temp\arduino\sketches\D5AFF887F9B4FFAFCC5B5151A01F0E99\sketch\sketch_nov8b.ino.cpp.o:sketch_nov8b.ino.cpp:(.text.HAL_TIM_PeriodElapsedCallback+0x0): first defined here
collect2.exe: error: ld returned 1 exit status

Uso la libreria SrcWrapper alla versione 1.0.1 nella cartella: C:\Users\MYSELF\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.6.0\libraries\SrcWrapper 
exit status 1

Compilation error: exit status 1

I need an interrupt to count timer pulses. This is just a simple example. I don't want to use HardwareTimer library to control timers because I want generate precise PWM.

**Comment by fpistm** [comment, marked as answer]
Hi @ennio64 
You have to disable the usage of HAL TIM for Arduino API:
https://github.com/stm32duino/Arduino_Core_STM32/wiki/HAL-configuration#hal-module-only

So define `HAL_TIM_MODULE_ONLY` thanks the build_opt.h or hal_conf_extra.h.
Then do not forget to add `extern "C"` before each handler definition, ex:
`extern "C" void TIM3_IRQHandler(void) {`

**Comment by ennio64** [comment]
SOLVED
I add "hal_conf_extra.h" (with #define HAL_TIM_MODULE_ONLY) in sketch folder ; add extern "C" before each handler definition and work.
Thanks.
